### ISSUE TITLE
Type checking does not pass for `lambda_stmt`

### ISSUE THREAD

**Issue by emontnemery** [sqlalchemy/sqlalchemy#9120]
### Describe the bug

The following does not pass type checking with latest mypy:

test.py:
```py
from __future__ import annotations
from sqlalchemy import lambda_stmt, Column, Integer, select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.sql.expression import StatementLambdaElement

class Base(DeclarativeBase):
    ...

class Events(Base):
    idx = Column(Integer)

def make_lambda() -> StatementLambdaElement:
    l = lambda: select(Events)
    reveal_type(l)
    reveal_type(l.__code__)
    reveal_type(l.__closure__)
    return lambda_stmt(lambda: select())
```

Type checking passes after simplifying `_LambdaType` from:
```py
class _LambdaType(Protocol):
    __code__: CodeType
    __closure__: Iterable[Tuple[Any, ...]]

    def __call__(self, *arg: Any, **kw: Any) -> ClauseElement:
        ...
```
To:
```py
class _LambdaType(Protocol):
    def __call__(self, *arg: Any, **kw: Any) -> ClauseElement:
        ...
```

Can `_LambdaType` be simplified like that, it doesn't seem necessary to type the functions `__code__` and `__closure__` dunders?

### To Reproduce

```python
> python3 -m venv .venv
> source .venv/bin/activate
> pip3 install sqlalchemy==2.0.0.rc3 sqlalchemy[mypy]==2.0.0.rc3 mypy==0.991
> mypy test.py
```

### Error

With the above example, mypy complains:
```
error: Argument 1 to "lambda_stmt" has incompatible type "Callable[[], Select[Any]]"; expected "_LambdaType"  [arg-type]
``` 

### Versions

- OS: Linux
- Python: 3.9
- SQLAlchemy: 2.0.0.rc3
- Database: N/A
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): N/A

**Comment by zzzeek** [comment]
still does not pass for pyright.    Will need to replace with `Callable[..., Any]`.  internal protocol stuff will have to be cast

**Comment by emontnemery** [comment]
It's a bit of a shame if the protocol is simplified to just a `Callable[..., Any]` type, does pyright not support protocols?

**Comment by zzzeek** [comment]
it does, it's just the protocol for "this must be a Python function or lambda and not an arbitrary callable" is not part of pep-484.

it's fine, looks like lambda.py needs annotations for common operations too.

**Comment by zzzeek** [comment]
lambda statements and sub-elements are currently not generically typed for their sub-element, so you get StatementLambdaElement, without a specific tuple return type mentioned.  typing will all pass, you just wont have `Result[Tuple[int, str, bool]]` for results, you'd have `Result[Any]`.   I think that is fine for now as lambda statements are a special case and you can use implicit casts to get what you need.

it would be hard to make specific expression and result typing work for lambdas without https://github.com/python/typing/issues/548, since it would need to be like `StatementLambdaElement[StatementType[ResultType]]`.

**Comment by zzzeek** [comment]
sure, return type should be ClauseElement, that's fine.

**Comment by zzzeek** [comment]
mmm, there's a few different kinds of lambdas but having them return something that's not Any makes mypy fail on all the lambdas, which would make the syntax require explicit annotations in all the lambdas, which is too verbose.   will start small here

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**implement basic typing for lambda elements** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4371

**Comment by emontnemery** [comment]
Will the fix be included in 2.0.0?
(I'm asking since I see other issues are tagged with 2.0.0)
